[native] Add Apache Arrow Flight Connector

BryanCutler opened the pull request:
## Description
<!---Describe your changes in detail-->
Add Prestissimo support for Apache Arrow Flight connectors.

The native Arrow Flight connector can be used to connect to any Arrow Flight enabled Data Source. The metadata layer is handled by the Presto coordinator and does not need to be re-implemented in C++. Any Java connector that inherits from `presto-base-arrow-flight` can use this connector as it's counterpart for the Prestissimo layer.

Different Arrow-Flight enabled data sources can differ in authentication styles. A plugin-style interface is provided to handle such cases with custom authentication code by extending `arrow_flight::auth::Authenticator`.

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

RFC: https://github.com/prestodb/rfcs/blob/main/RFC-0004-arrow-flight-connector.md#prestissimo-implementation

Continues from https://github.com/prestodb/presto/pull/24082

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->
Arrow Flight based connector will be supported in Prestissimo.

## Test Plan
<!---Please fill in how you tested your change-->
Unit Tests set up a testing Arrow Flight server and exchange data using the new connector.

## Contributor checklist

- [x] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Prestissimo (Native Execution)
* Add support for Apache Arrow Flight connectors
```

steveburnett commented:
Thanks for the release note!

New release note guidelines. Please remove the manual PR link in the following format from the release note entries for this PR.

```

:pr:`**`

```

I have updated the [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) to remove the examples of manually adding the PR link.

BryanCutler commented:
Great work from @Rijin-N in https://github.com/prestodb/presto/pull/24082. I didn't make any major changes, just updated the protocol to what was merged in Java from https://github.com/prestodb/presto/pull/24427, fixed some build issues, added tests for custom authentication and some minor tweaks. I'll try to highlight my changes below.

steveburnett commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/data/README.md:
```diff
@@ -0,0 +1,6 @@
+The `tls_certs` directory contains placeholder TLS certificates generated for unit testing the Arrow Flight Connector with TLS enabled. These certificates are not intended for production use and should only be used in the context of unit tests.
+
+### Generating TLS Certificates
+To create the TLS certificates and keys inside the `tls_certs` folder, run the following command:
```
```suggestion
To create the TLS certificates and keys inside the `tls_certs` folder, run the following command:

```
Add a line after this blank line to separate this text and the command. Without the line, it looks like this: 

<img width="841" alt="Screenshot 2025-02-05 at 4 51 27 PM" src="https://github.com/user-attachments/assets/b086ec88-7b14-4375-9c97-9c439bdcf12b" />

steveburnett reviewed (changes requested):
Thanks for the doc! One nit of formatting.

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.cpp:
```diff
@@ -0,0 +1,184 @@
+  }
+}
+
+void FlightDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
```
Did some minor cleanup in this method to simply use of the serialized FlightEndpoint

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+# limitations under the License.
+find_package(Arrow REQUIRED)
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(ARROW_FLIGHT REQUIRED IMPORTED_TARGET GLOBAL arrow-flight)
```
I needed to add `GLOBAL` for using PkgConfig to prevent an `-larrow-flight` linker error. This seems like a better general solution to reference the libraries, see https://stackoverflow.com/questions/29191855/what-is-the-proper-way-to-use-pkg-config-from-cmake for more details

majetideepak replied:
Why choose PkgConfig approach and not how Velox finds Arrow? We should use one approach for consistency.

BryanCutler replied:
@Rijin-N had it building with PkgConfig, and one of the first things I tried was changing this to the standard approach. I could not get it to find the flight libraries in my environment though. The only way I could get it to work was to revert back to PkgConfig and add `GLOBAL`.

@Rijin-N do you remember why PkgConfig was used?

Rijin-N replied:
@BryanCutler I also tried to use the velox way of finding arrow-flight libraries. But it did not work. Thats why we decided to use `PkgConfig`

BryanCutler replied:
@majetideepak I went back and tried this again, using find_package like this in `arrow_flight/CMakeLists.txt`
```
find_package(Arrow REQUIRED)
find_package(ArrowFlight REQUIRED)
```

I get this CMake error:
```
CMake Error at ../working/deps-install/lib64/cmake/ArrowFlight/ArrowFlightConfig.cmake:64 (arrow_keep_backward_compatibility):
  Unknown CMake command "arrow_keep_backward_compatibility".
Call Stack (most recent call first):
  presto_cpp/main/connectors/arrow_flight/CMakeLists.txt:13 (find_package)
```
This errors on the line to `find_package(ArrowFlight Required)` but `arrow_keep_backward_compatibility` is defined in ArrowConfig.cmake and I'm not sure why it can't find it, but this isn't a problem when using PkgConfig.

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/FlightConfig.cpp:
```diff
@@ -0,0 +1,45 @@
+  return config_->get<bool>(kServerVerify, true);
+}
+
+std::optional<std::string> FlightConfig::serverSslCertificate() {
```
This was `folly::Optional`, changed to be consistent

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorAuthTest.cpp:
```diff
@@ -0,0 +1,251 @@
+/*
```
Added these tests to verify adding a custom authenticator works

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+
+target_link_libraries(
+  presto_flight_connector_infra_test presto_protocol
+  presto_flight_connector_test_lib GTest::gtest GTest::gtest_main ${GLOG})
```
Added G stuff due to linker error

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/FlightConnectorTestBase.h:
```diff
@@ -0,0 +1,88 @@
+  virtual arrow::flight::Location getServerLocation();
+
+  virtual void setFlightServerOptions(
+      arrow::flight::FlightServerOptions* serverOptions) {}
```
Server options now specified by subclasses

steveburnett reviewed (approved):
LGTM! (docs)

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/auth/Authenticator.cpp:
```diff
@@ -0,0 +1,48 @@
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::presto::connector::arrow_flight::auth {
+namespace {
```
changes this to be in an anonymous namespace, instead of static

BryanCutler commented on presto-native-execution/presto_cpp/main/SystemConnector.cpp:
```diff
@@ -351,7 +351,8 @@ std::unique_ptr<velox::connector::ConnectorSplit>
     const protocol::ConnectorSplit* const connectorSplit,
-    const protocol::SplitContext* splitContext) const {
+    const protocol::SplitContext* splitContext,
+    const std::map<std::string, std::string>& extraCredentials) const {
```
@Rijin-N @elbinpallimalilibm is there a specific requirement to use these `extraCredentials` to authenticate a Flight client? I would think most users would prefer to authenticate the client with a token in the header, rather than use this. The Java Arrow Connector does not have this either.

ashkrisk replied:
@BryanCutler the base Presto (Java) Flight Connector is a template and not a full connector, and elects not to do authentication at all. The idea is that the authentication mechanism is largely dependent on the specific Flight server, so the whole business of authentication should be dealt with by the specialized connectors.

The problem we need to solve is that different Presto users may have different sets of access permissions on the connected Arrow Flight server. In this case we need to "pass through" the Presto user's authentication to the connected Flight Server, which means we somehow need to send the credentials from the Presto coordinator to the Presto workers.

In Java, there is an Identity API which we can use through `connectorSession.getIdentity().getExtraCredentials()`.
IBM's specialization of the base Flight connector uses this API to pass the user's token to the workers. This isn't possible in Prestissimo because Velox connectors have no concept of Identity at all, so we need to pass the token in the split. 

This issue provides more context:
https://github.com/prestodb/presto/issues/22849

There's also a corresponding Velox issue:
https://github.com/facebookincubator/velox/issues/10107

ashkrisk replied:
@BryanCutler we need this to get the multi-user setup working (IBM's flavor of the Flight connector relies on it). You can leave this off for now as long as you're able to maintain the diff in IBM code until there's a better solution in OSS.

BryanCutler replied:
@ashkrisk we need to make sure we have a discussion and proper design before adding APIs like this, since this also affects other connectors. I believe it was mentioned there could be a security concern if the credentials are added to the split, which is then serialized.

ethanyzhang replied:
@BryanCutler do we need to get a meeting to talk about this?

BryanCutler replied:
I discussed with @majetideepak and @aditi-pandit , and have a possible way forward that I'm looking into, and I'll report back

ashkrisk replied:
@BryanCutler With regards to security, any method used to transfer the credentials from the coordinator to the worker requires serialization at some point. This includes the `Identity` framework available in Java. To ensure security, you need to ensure encrypted connections between all Presto nodes.

BryanCutler replied:
The plan is to incorporate https://github.com/prestodb/presto/pull/22859 which will add the extraCredentials to the connector session properties. This will be available to the Flight data source when it is processing the split and can use them for authentication. I'll try to add a test for this if I'm able to.

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/FlightConfig.cpp:
```diff
@@ -0,0 +1,45 @@
+}
+
+bool FlightConfig::serverVerify() {
+  return config_->get<bool>(kServerVerify, false);
```
Changed default to "false" to be consistent with Java

elbinpallimalilibm replied:
server verify is by default true in Java
<img width="708" alt="image" src="https://github.com/user-attachments/assets/5a1846d1-4fb1-4d60-878e-9698f2618cec" />

BryanCutler replied:
Good point, the default behavior for the Flight client is to raise an error if using TLS and a certificate is provided but verify is disabled, and you need to explicitly disable it when using TLS.

I didn't catch that the Java connector defaults the config to NULL, which translates to true. I think that's confusing and we should not make it nullable. I'll change Java and here to default to true.

BryanCutler replied:
updated the Java side in https://github.com/prestodb/presto/pull/24518 and changed the default here too

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorTlsTest.cpp:
```diff
@@ -0,0 +1,138 @@
+                {1, 12, 2, std::numeric_limits<int64_t>::max()})}));
+
+    auto idVec = makeFlatVector<int64_t>(
+        {1, 12, 2, std::numeric_limits<int64_t>::max()});
```
Abstract a variable for {1, 12, 2, std::numeric_limits<int64_t>::max()} so that it can be used in the input and result vector.

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorTlsTest.cpp:
```diff
@@ -0,0 +1,138 @@
+
+namespace {
+
+namespace velox = facebook::velox;
```
Haven't seen this style of shorthand for namespaces in Prestissimo code. Can these be removed ?

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorTest.cpp:
```diff
@@ -0,0 +1,197 @@
+      "missing columnHandle for column 'value'");
+}
+
+TEST_F(FlightConnectorTest, dataSourceTest) {
```
test names shouldn't have the word "Test" in them. Please fix the names.

aditi-pandit reviewed (commented):
Thanks @BryanCutler. Did a quick round of the code.

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/ConnectorRegistration.cpp:
```diff
@@ -0,0 +1,40 @@
+
+namespace facebook::presto::connector {
+
+void registerAllPrestoConnectors() {
```
The file names can also be `Registration.h/.cpp` since they are in the connectors folder.
`presto-native-execution/presto_cpp/main/connectors/Registration.h`

majetideepak replied:
Let's move the SystemConnector here as well.

BryanCutler replied:
ok sounds good

BryanCutler replied:
@majetideepak should SystemConnector then be in `namespace facebook::presto::connector` like the ArrowFlightConnector?

BryanCutler replied:
> Let's move the SystemConnector here as well.

Could you clarify if you mean the source files SystemConnector.cpp/h or just the registration?

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/Macros.h:
```diff
@@ -0,0 +1,50 @@
+  } while (false)
+
+#define AFC_ASSIGN_OR_RAISE_IMPL(result_name, lhs, rexpr)            \
+  auto&& result_name = (rexpr);                                      \
```
I don't think that will work, the `lhs` variable gets scoped in the loop. So something like this you can't do anymore
```
AFC_ASSIGN_OR_RAISE(auto client, FlightClient::Connect(loc, *clientOpts_));
```
I believe this was taken from https://github.com/facebookincubator/velox/blob/d3f7b3d8c5c67bdcdcad05d9946ae6b87f88d760/velox/functions/prestosql/json/SIMDJsonUtil.h#L36
Which doesn't wrap in a loop either

ashkrisk replied:
Actually, it comes from the Arrow source:
https://github.com/apache/arrow/blob/152eb879b4fd4476dadd3cd497bb8d78d3013cef/cpp/src/arrow/result.h#L460

This macro declares and assigns to a variable. If wrapped in a loop, the assignment is scoped to the loop and variable is unavailable outside, essentially making the macro useless.

See also the comments around `ARROW_ASSIGN_OR_RAISE` in the source:
https://github.com/apache/arrow/blob/152eb879b4fd4476dadd3cd497bb8d78d3013cef/cpp/src/arrow/result.h#L467-L487

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/TestFlightServerTest.cpp:
```diff
@@ -0,0 +1,84 @@
+
+std::unique_ptr<TestFlightServer> TestFlightServerTest::server;
+
+TEST_F(TestFlightServerTest, basicTest) {
```
I'm not too sure either, it's part of `presto_flight_connector_infra_test` tests, which is separate. It starts a server, adds data, then makes it's own client to get the flight stream - not using the connector. So it seems to be testing that the test server can serve Flight clients correctly, but that is also covered with the regular connector tests. @Rijin-N could you comment on this test?

ashkrisk replied:
The goal here is to run a simple test to be confident that the Test Flight Server is working correctly. Now, any failures that happen in the actual connector tests can be safely assumed to be a problem with the connector and not with the server, potentially making things easier to debug.

This is like the difference between testing each component vs having only integration tests.

Rijin-N commented on presto-native-execution/scripts/setup-adapters.sh:
```diff
@@ -35,15 +35,73 @@ function install_prometheus_cpp {
+  wget_and_untar https://github.com/apache/arrow/archive/apache-arrow-${ARROW_VERSION}.tar.gz arrow
+  cmake_install_dir arrow/cpp \
+    -DARROW_FLIGHT=ON \
+    -DARROW_BUILD_BENCHMARKS=ON \
```
Hi @BryanCutler @majetideepak,

This was included due to a known issue in Arrow version 15.0.0. Without setting ARROW_BUILD_BENCHMARKS=ON, a compilation error occurs while building the Arrow package. This issue has been fixed in later versions of Arrow.

Please see the PR link for details:
PR: https://github.com/apache/arrow/pull/41622/files

majetideepak replied:
Thanks! Let's add a comment here about this so that this can be removed when Arrow is updated.

majetideepak commented on presto-native-execution/scripts/setup-adapters.sh:
```diff
@@ -35,15 +35,73 @@ function install_prometheus_cpp {
+}
+
+function install_arrow_flight {
+  ARROW_VERSION="15.0.0"
```
We should get this from the Velox script. We can skip here to align the versions.

BryanCutler replied:
@majetideepak the version is defined in the OS setup scripts, we could do something like
```
ARROW_VERSION="${ARROW_VERSION:-15.0.0}"
```

BryanCutler commented on presto-native-execution/scripts/setup-adapters.sh:
```diff
@@ -35,15 +35,73 @@ function install_prometheus_cpp {
   cmake_install -DBUILD_SHARED_LIBS=ON -DENABLE_PUSH=OFF -DENABLE_COMPRESSION=OFF
 }
 
+function install_abseil {
```
@majetideepak they are the same versions, but they are done in the call to `install_gcs-sdk-cpp` so we can't call them directly unless we update the velox script.

majetideepak replied:
I will fix this on the Velox side to make it reusable.

BryanCutler commented:
@aditi-pandit @majetideepak I've addressed most of the feedback, just had a couple questions above. The connectors have been moved, namespaces cleaned and files named consistently. Please take another look, thanks!

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConfig.cpp:
```diff
@@ -0,0 +1,45 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "presto_cpp/main/connectors/arrow_flight/ArrowFlightConfig.h"
```
nit: new line above.

majetideepak replied:
I don't think we need an extra `arrow_flight` directory. All the files have ArrowFlight prefix anyway.

BryanCutler replied:
There are a quite a few files for the arrow flight connector, not all of them have the prefix. For example the "auth" stuff is only for the Arrow connector, Utils for Arrow testing, and tls certs used for testing. I think it makes sense to keep all this in a directory otherwise it will be a mess for any other connectors added.

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConfig.h:
```diff
@@ -0,0 +1,57 @@
+  static constexpr const char* kServerSslCertificate =
+      "arrow-flight.server-ssl-certificate";
+
+  std::string authenticatorName();
```
std::string authenticatorName() const; here and below.

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConfig.h:
```diff
@@ -0,0 +1,57 @@
+  std::optional<std::string> serverSslCertificate();
+
+ private:
+  std::shared_ptr<const velox::config::ConfigBase> config_;
```
const

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.h:
```diff
@@ -0,0 +1,194 @@
+  void addDynamicFilter(
+      velox::column_index_t outputChannel,
+      const std::shared_ptr<velox::common::Filter>& filter) override {
+    VELOX_NYI("This connector doesn't support dynamic filters");
```
VELOX_UNSUPPORTED("Arrow Flight connector doesn't support dynamic filters");

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.h:
```diff
@@ -0,0 +1,194 @@
+ */
+#pragma once
+
+#include <arrow/flight/api.h>
```
Let's use the PIMPL idiom and limit arrow headers to the cpp file.
See example usage here https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.h#L77

BryanCutler replied:
I was able to move things around a bit and just forward declare the arrow classes to get rid of the arrow includes. They are still in a couple of the test headers, though. That's fine right?

majetideepak replied:
test headers are fine.

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/auth/Authenticator.h:
```diff
@@ -0,0 +1,85 @@
+ */
+#pragma once
+
+#include <arrow/flight/api.h>
```
We should avoid exposing this arrow header as well.
Can we use PIMPL here as well?

majetideepak replied:
No need of PIMPL. Forward declare and make API
```
virtual void authenticateClient(
      arrow::flight::FlightClient* client,
      const velox::config::ConfigBase* sessionProperties,
      arrow::flight::AddCallHeaders* headerWriter) = 0;
```

BryanCutler commented on presto-native-execution/presto_cpp/main/PrestoServer.cpp:
```diff
@@ -254,33 +252,9 @@ void PrestoServer::run() {
-      std::make_unique<SystemPrestoToVeloxConnector>("$system@system"));
+
+  // Register Presto connector factories and connectors
+  presto::registerConnectors();
```
we need it, this file isn't using namespace presto

majetideepak replied:
I see `namespace facebook::presto {` on line 82?

BryanCutler replied:
Oh right, actually there is already a method `PrestoServer::registerConnectors` I guess that was causing a conflict. It's kind of confusing to have both with the same name. The one in PrestoServer gets the connector and calls `velox::connector::registerConnector(connector);`

Maybe we could go back to calling the new one `registerPrestoConnectors`?

majetideepak replied:
That should probably be renamed as `registerVeloxConnectors`.

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorAuthTest.cpp:
```diff
@@ -0,0 +1,236 @@
+  static constexpr const char* kAuthToken = "Bearer 1234";
+  static constexpr const char* kAuthTokenUnauthorized = "Bearer 2112";
+
+  arrow::Status StartCall(
```
`startCall`

BryanCutler replied:
this overrides from `flight::ServerMiddlewareFactory`

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/ArrowFlightConnectorTestBase.h:
```diff
@@ -0,0 +1,89 @@
+/// tests should call FlightWithServerTestBase::updateTables to populate it.
+class FlightWithServerTestBase : public ArrowFlightConnectorTestBase {
+ public:
+  static constexpr const char* BIND_HOST = "127.0.0.1";
```
kBindHost

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/ArrowFlightConnectorTestBase.h:
```diff
@@ -0,0 +1,89 @@
+class FlightWithServerTestBase : public ArrowFlightConnectorTestBase {
+ public:
+  static constexpr const char* BIND_HOST = "127.0.0.1";
+  static constexpr const char* CONNECT_HOST = "localhost";
```
kConnectHost

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/ArrowFlightConnectorTestBase.h:
```diff
@@ -0,0 +1,89 @@
+ public:
+  static constexpr const char* BIND_HOST = "127.0.0.1";
+  static constexpr const char* CONNECT_HOST = "localhost";
+  constexpr static int LISTEN_PORT = 5000;
```
kListenPort

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/ArrowFlightPlanBuilder.cpp:
```diff
@@ -0,0 +1,45 @@
+
+namespace facebook::presto::test {
+namespace {
+static const std::string kFlightConnectorId = "test-flight";
```
can remove static since its inside an anonymous namespace

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/TestingArrowFlightServer.h:
```diff
@@ -0,0 +1,48 @@
+    tables_.erase(name);
+  }
+
+  arrow::Status DoGet(
```
doGet

BryanCutler replied:
This one overrides `arrow::flight::FlightServerBase`

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/Utils.cpp:
```diff
@@ -0,0 +1,90 @@
+      arrow::Table::FromRecordBatches({makeRecordBatch(names, arrays)}));
+}
+
+std::string readFile(const std::string& path) {
```
Is this used anywhere?

majetideepak reviewed (commented):
Some more comments.

BryanCutler commented:
E2E tests with Java Presto have been added in the last commit, they use the ArrowFlightQueryRunner

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorDataTypeTest.cpp:
```diff
@@ -0,0 +1,328 @@
+
+class ArrowFlightConnectorDataTypeTest : public FlightWithServerTestBase {};
+
+TEST_F(ArrowFlightConnectorDataTypeTest, booleanType) {
```
@BryanCutler : Is it possible to exchange complex types ? Can you add a test if we can.

BryanCutler replied:
sure, will try

BryanCutler replied:
added tests for array, map and row types

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorDataTypeTest.cpp:
```diff
@@ -0,0 +1,328 @@
+      .assertResults(makeRowVector({realVec, doubleVec}));
+}
+
+TEST_F(ArrowFlightConnectorDataTypeTest, varcharType) {
```
Please add tests with length > 12 since that is out of line. Also add tests with special characters and varbinary.

BryanCutler replied:
What is the issue with length > 12?

BryanCutler replied:
ah, I see.. > 12 then a reference to the string is held

aditi-pandit reviewed (commented):
@BryanCutler : Few more comments.

BryanCutler commented on .github/workflows/arrow-flight-tests.yml:
```diff
@@ -48,6 +48,142 @@ jobs:
-        run: ./mvnw test ${MAVEN_TEST} -pl ${{ matrix.modules }}
+        run: ./mvnw test ${MAVEN_TEST} -pl ${{ matrix.modules }} -Dtest="*,!TestArrowFlightNativeQueries"
+
+  prestocpp-linux-build-for-test:
```
@majetideepak @czentgr I took most of this from `prestocpp-linux-build-and-unit-test.yml`, but it seems to be timing out waiting for the worker nodes to come up and I'm not sure why. Do you have any ideas to troubleshoot?

BryanCutler commented:
@aditi-pandit and @majetideepak I've addressed the latest feedback, more data type tests have been added. I'm still trying to troubleshoot the e2e ci testing with Java, but please review again.

majetideepak commented:
@BryanCutler Add the installs needed as a CI step for now. We will include that in the image after merge and remove the CI step.

BryanCutler commented:
@majetideepak @aditi-pandit I've addressed the feedback, worked out the e2e tests in CI and all are passing now, please revew

BryanCutler commented on .github/workflows/arrow-flight-tests.yml:
```diff
@@ -48,6 +48,161 @@ jobs:
+            -DTREAT_WARNINGS_AS_ERRORS=1 \
+            -DENABLE_ALL_WARNINGS=1 \
+            -DVELOX_ENABLE_ARROW=OFF \
+            -DVELOX_ENABLE_PARQUET=OFF \
```
NOTE: enabling parquet seemed to cause a build issue, need to followup with some more investigation

BryanCutler commented on presto-native-execution/presto_cpp/main/PrestoServer.cpp:
```diff
@@ -21,11 +21,12 @@
 #include "presto_cpp/main/common/Counters.h"
 #include "presto_cpp/main/common/Utils.h"
+#include "presto_cpp/main/connectors/Registration.h"
+#include "presto_cpp/main/connectors/SystemConnector.h"
```
It's still needed when the system connector is registered with velox here https://github.com/prestodb/presto/blob/169382b2ae2431d95dc419077a58270c798fa001/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1221

My feeling is there is still some needed refactoring that needs to be done with connector registration. I would recommend we don't do that here, since this is already a large PR.

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/auth/Authenticator.h:
```diff
@@ -0,0 +1,89 @@
+
+#include "velox/common/config/Config.h"
+
+namespace arrow::flight {
```
That was suggested by @majetideepak https://github.com/prestodb/presto/pull/24504#discussion_r1968444534
Trying to avoid the Arrow includes from Presto headers

aditi-pandit replied:
@majetideepak : In past work we definitely followed this approach to minimize the headers, but I didn't see it in Velox/Presto Native code. Any particular reason that you have suggested it ?

I'm also confused about calling this file a Presto header... Will this be included in Presto code without arrow dependencies ?

BryanCutler replied:
One of the benefits of forward declaring here is that if you are developing a plugin authentication and you only care about checking a user/password from the sessionProperties, then you wouldn't need to link against the Arrow Flight libraries to build the plugin.

Generally speaking, forward declaration would also help to minimize conflicts with definitions or declarations in other third party libraries by not bringing all libraries into scope together. So in `Registration.cpp`, where all connector headers are brought in, we forward the Arrow classes instead of including the Arrow headers and there is no chance of conflicts.

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.h:
```diff
@@ -0,0 +1,184 @@
+#include "presto_cpp/main/connectors/arrow_flight/auth/Authenticator.h"
+#include "velox/connectors/Connector.h"
+
+namespace arrow {
```
Suggested by @majetideepak https://github.com/prestodb/presto/pull/24504#discussion_r1968436851 to avoid Arrow includes in the headers

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/TestingArrowFlightServer.cpp:
```diff
@@ -0,0 +1,35 @@
+  }
+  auto& table = it->second;
+  auto reader = std::make_shared<arrow::TableBatchReader>(table);
+  *stream = std::make_unique<RecordBatchStream>(std::move(reader));
```
I don't think the move is necessary, but having separate lines improves readability for this testing utility class. I'll remove the `std::move`

BryanCutler commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/TestingArrowFlightServer.h:
```diff
@@ -0,0 +1,48 @@
+ public:
+  TestingArrowFlightServer() = default;
+
+  void updateTable(std::string name, std::shared_ptr<arrow::Table> table) {
```
I don't think transfer of ownership is needed, the Table will be immutable, so I can change to store a const ptr. No need to make it overly complicated.

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/Utils.h:
```diff
@@ -0,0 +1,66 @@
+namespace facebook::presto::test {
+
+template <typename T>
+std::shared_ptr<arrow::Array> makeNumericArray(
```
Nit : Add an alias for std::shared_ptr<arrow::Array> to reuse 

using ArrowArrayPtr = std::shared_ptr<const arrow::Array>;

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/Utils.cpp:
```diff
@@ -0,0 +1,108 @@
+    const arrow::ArrayVector& arrays) {
+  VELOX_CHECK_EQ(names.size(), arrays.size());
+
+  auto nrows = (!arrays.empty()) ? (arrays[0]->length()) : 0;
```
Nit : Rename variable to numRows.

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/utils/ArrowFlightConnectorTestBase.h:
```diff
@@ -0,0 +1,95 @@
+
+static const std::string kFlightConnectorId = "test-flight";
+
+class ArrowFlightConnectorTestBase
```
Why is this class abstracted ? This class can be combined with the child class below.

BryanCutler replied:
I think it was done to have a test base class that doesn't startup a server, but it's not used so I'll combine them

ashkrisk replied:
Note that in IBM code we use this class directly to run an integration test against an actual Flight Server - this would be more difficult if the two classes are combined.

BryanCutler replied:
@ashkrisk we can't add additional classes in open-source for the only purpose to be internal testing. The base class without the server only registers the connector, that could easily be done in the internal integration tests.

aditi-pandit replied:
+1.

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/tests/ArrowFlightConnectorAuthTest.cpp:
```diff
@@ -0,0 +1,236 @@
+  std::mutex mutex_;
+};
+
+class TestingAuthenticator : public Authenticator {
```
Nit : Names TestAuthenticator and TestAuthenticatorFactory are better than usage of verbs like Testing in the name.

aditi-pandit reviewed (commented):
@BryanCutler : The code is looking good. I have few minor comments. But overall I'm satisfied with the code. Will give an approval once Deepak approves all the build related changes.

majetideepak commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightNativeQueries.java:
```diff
@@ -0,0 +1,427 @@
+
+    public static int getServerPort()
+    {
+        return 9443;
```
We should use PortUtil `velox/exec/tests/utils/PortUtil.h` to get a random port and avoid test failure due to the port being unavailable.

BryanCutler replied:
PortUtil is only in C++, I updated with a Java equivalent here and will go back and try to use a random port with PortUtil in the native arrow connector tests

BryanCutler replied:
both native and additional java tests are now using a random free port

majetideepak commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightNativeQueries.java:
```diff
@@ -0,0 +1,427 @@
+        assertQuery("SELECT * FROM nation WHERE nationkey NOT IN (1, 2, 3)");
+        assertQuery("SELECT * FROM nation WHERE nationkey NOT IN (-14, 2)");
+        assertQuery("SELECT * FROM nation WHERE nationkey NOT IN (1, 2, 3, 4, 5, 10, 11, 12, 13)");
+        // Java coordinator/workers causes these queries to fail, even though the INT_MAX ones work on cpp
```
Can you elaborate this?

BryanCutler replied:
I copied some tests from AbstractTestNativeGeneralQueries, I left the test as it was, including this comment, which wasn't from me https://github.ibm.com/lakehouse/presto/blob/719bfe2a4d35703e04b3127f6db9d37b94e8dc8c/presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeGeneralQueries.java#L161

I'm not exactly sure why they fail for Java either, I can remove the commented out tests though, since it is a copy

majetideepak commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightQueries.java:
```diff
@@ -83,11 +83,16 @@ public void close()
 
+    public static int getServerPort()
+    {
+        return 9443;
```
Use PortUtil

BryanCutler replied:
Updated native to use PortUtil to get a free port

majetideepak commented on presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.cpp:
```diff
@@ -0,0 +1,209 @@
+#include "presto_cpp/main/connectors/arrow_flight/Macros.h"
+#include "velox/vector/arrow/Bridge.h"
+
+using namespace arrow::flight;
```
Can we avoid this and explicitly use the namespace where needed?

BryanCutler commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightEchoQueries.java:
```diff
@@ -124,11 +124,10 @@ public void setup()
 
         allocator = new RootAllocator(Long.MAX_VALUE);
-
+        File certChainFile = new File("src/test/resources/server.crt");
```
No, I was just trying to make the setup consistent across the arrow tests, I can revert though to keep changes to a min

BryanCutler commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightIntegrationSmokeTest.java:
```diff
@@ -65,8 +65,8 @@ protected QueryRunner createQueryRunner()
+        arrowFlightQueryRunner.close();
         server.close();
         allocator.close();
-        arrowFlightQueryRunner.close();
```
This is needed to ensure the correct order of closing

BryanCutler replied:
although it's unrelated to native testing. I'll revert this too and I'll open another PR for some of these small fixes.

aditi-pandit commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightNativeQueries.java:
```diff
@@ -0,0 +1,427 @@
+/*
```
@czentgr : I gave this a thought as well ... But then felt that having native as well as java connector tests in the connector module would actually be a good approach.  Both implementations cohesively work and test together.

Are there particular reservations about this ?

BryanCutler replied:
All of the Java Flight stuff is under the presto-base-arrow-flight module and only tested as part of the arrow-flight-tests workflow. I also thought about adding the native tests to presto-native-execution but there are a lot of things that would be needed from presto-base-arrow-flight which adds a lot to presto-native-execution for just a few tests. So it seemed like it was better to keep in compartmentalized in it's own module.

BryanCutler commented on .github/workflows/arrow-flight-tests.yml:
```diff
@@ -48,6 +48,142 @@ jobs:
+            -B _build/release \
+            -GNinja \
+            -DTREAT_WARNINGS_AS_ERRORS=1 \
+            -DENABLE_ALL_WARNINGS=1 \
```
> Arrow flight is behind the flag but it isn't turned on here 

I do have the flag `-DPRESTO_ENABLE_ARROW_FLIGHT_CONNECTOR=ON` enabled here

> rename the test java file TestPrestoNativeArrowFlightQueries.java. Then it is picked up automatically.

This is only if the test is moved to `presto-native-execution` right?

majetideepak reviewed (approved):
All my comments have been addressed. Thanks @BryanCutler

aditi-pandit commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightNativeQueries.java:
```diff
@@ -0,0 +1,387 @@
+        }
+    }
+
+    public static Map<String, String> getNativeWorkerSystemProperties()
```
@BryanCutler : Can we use the functions from PrestoNativeQueryRunnerUtils for getting the properties and external worker launcher instead of repeating the code here ?

That would mean adding a dependency on presto-native-execution, but that should be fine right ?
